virt-controller: Decouple network annotations generation

orelmisan opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Consider creating this PR as draft: https://github.com/kubevirt/kubevirt/blob/main/CONTRIBUTING.md#consider-opening-your-pull-request-as-draft
2. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

### What this PR does
Currently, the logic for generating virt-launcher pod's annotations is located in a central place under the TemplateService.

Annotations are tested in `template_test.go` file, which currently has more than 5K lines.

In an effort to decouple sig-network's logic from the core logic, introduce a mechanism which enables SIGs
to decouple their annotations generation logic from the core logic.

Doing so has several advantages:
1. Making the core logic more concise and less prune to change.
2. Making template_test.go shorter and focused on core functionality.
3. Allowing SIGs to rigorously test their logic, unburdened by
TemplateService's dependencies.
4. Allowing autonomy for SIG approvers to own their logic.

Introduce sig-network's annotations generator, move relevant logic to it.

<!-- (optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*: -->
Fixes #

### Why we need it and why it was done in this way
The following tradeoffs were made:

The following alternatives were considered:

Links to places where the discussion took place: <!-- optional: slack, other GH issue, mailinglist, ... -->

### Special notes for your reviewer

<!-- optional -->

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [X] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [X] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
None
```

orelmisan commented:
/uncc @AlonaKaplan @stu-gott

orelmisan commented:
/sig code-quality

orelmisan commented:
[Change](https://github.com/kubevirt/kubevirt/compare/82bbb0a3fe08a4dc4c9ec7ee2e319cbd6ec5d8bc..b449484d10b791d51914e0175f2846eb9f5d8d32): Rebase.

orelmisan commented:
/cc @ormergi

orelmisan commented:
[Change](https://github.com/kubevirt/kubevirt/compare/fb16c22b68b40b321c4ee045cfaed34fe96d29f2..83ff444e7fc8566d83eb13c055aa220fc38f4f0a): Rebase.

EdDev commented on pkg/network/pod/annotations/generator_test.go:
```diff
@@ -135,6 +136,39 @@ var _ = Describe("Annotations Generator", func() {
+			Expect(annotations).To(Not(HaveKey(istio.KubeVirtTrafficAnnotation)))
+		},
+			Entry("when VMI is not connected to pod network", libvmi.New()),
+			Entry("when VMI is connected to pod network and not using masquerade binding",
```
An unrelated topic for a follow up discussion.

The means that for passt, we do not add a special annotation.
And I think ISTIO still works, so I would be interested to get to the bottom of this: Why ISTIO needs it and how it related to Kubevirt. I suspect ISTIO adds some nftables rules to integrate with the ones we add to the masquerade binding, but I am unsure.

@maiqueb , @qinqon , FYI

EdDev reviewed (commented):
Thank you for the decoupling work!

I reached the 3rd commit so far.
One more to go.

>  virt-ctrl: Extract Istio annotation generation
> 
> This commits keeps the existing logic that adds
> the said annotation in case there is any interface
> with masquerade binding.
> It could be adjusted in a follow-up PR.

The text is a bit confusing.
I do not understand what it does not do that requires a follow up.
Not that I think it merits an explanation about what is not done.. only about a more accurate explanation of what is moved.

orelmisan commented on pkg/virt-controller/services/template.go:
```diff
@@ -1367,6 +1372,15 @@ func generatePodAnnotations(vmi *v1.VirtualMachineInstance, config *virtconfig.C
 	annotationsSet[descheduler.EvictOnlyAnnotation] = ""
 
+	for _, generator := range t.annotationsGenerators {
+		generatedAnnotations, err := generator.Generate(vmi)
```
Done.

orelmisan commented on pkg/virt-controller/services/template.go:
```diff
@@ -1541,3 +1555,9 @@ func WithNetBindingPluginMemoryCalculator(netBindingPluginMemoryCalculator netBi
 	}
 }
+
+func WithAnnotationsGenerator(annotationsGenerator annotationsGenerator) templateServiceOption {
```
Done.

orelmisan commented on pkg/virt-controller/services/template_test.go:
```diff
@@ -5109,10 +4983,10 @@ var _ = Describe("Template", func() {
 				generator2Value = "generator2Value"
 			)
-			testAnnotationGenerator1 := &stubAnnotationGenerator{
+			testAnnotationsGenerator1 := &stubAnnotationsGenerator{
```
Sorry for that.
Done.

orelmisan commented on pkg/virt-controller/services/template_test.go:
```diff
@@ -5100,6 +5100,50 @@ var _ = Describe("Template", func() {
 		})
 	})
+
+	Context("Custom annotations Generation", func() {
```
> It seems some logic is not checked:
> 
> * What happens on failure to generate a set of annotations?
> * A generator has more than one key/value.
> * The generated annotation will override an existing one (e.g. the 2nd generator overrides the 1st output).

A test case was added for each proposed scenario.

> The last point opens up the question if it should fail with an error or just emit a warning or ignore as it is implemented now.

Currently the last generator will silently override any shared key's value.
It has some risk, but IMO it is acceptable for now.

orelmisan commented on pkg/network/pod/annotations/generator.go:
```diff
@@ -0,0 +1,66 @@
+}
+
+func (g Generator) Generate(vmi *v1.VirtualMachineInstance) (map[string]string, error) {
+	annotations := map[string]string{}
```
Done.

orelmisan commented on pkg/network/pod/annotations/generator_test.go:
```diff
@@ -0,0 +1,155 @@
+			kubevirtNamespace = "kubevirt"
+		)
+
+		const (
```
The two groups are now united.

orelmisan replied:
Eventually I went with another approach that limits constants' scope.

orelmisan commented on pkg/network/pod/annotations/generator_test.go:
```diff
@@ -0,0 +1,155 @@
+		})
+
+		It("should generate the Multus networks annotation", func() {
+			generator := annotations.NewGenerator(clusterConfig)
```
Done.

orelmisan commented on pkg/network/pod/annotations/generator.go:
```diff
@@ -62,5 +63,17 @@ func (g Generator) Generate(vmi *v1.VirtualMachineInstance) (map[string]string,
 	}
 
+	if shouldAddIstioKubeVirtAnnotation(vmi) {
+		annotations[istio.KubeVirtTrafficAnnotation] = "k6t-eth0"
```
Done.

orelmisan commented on pkg/network/pod/annotations/generator_test.go:
```diff
@@ -135,6 +136,39 @@ var _ = Describe("Annotations Generator", func() {
+
+			Expect(annotations).To(Not(HaveKey(istio.KubeVirtTrafficAnnotation)))
+		},
+			Entry("when VMI is not connected to pod network", libvmi.New()),
```
Changed the test's title to "when VMI is not connected any network".

orelmisan commented:
[Change](https://github.com/kubevirt/kubevirt/compare/83ff444e7fc8566d83eb13c055aa220fc38f4f0a..8b0d3c8de4684f7c0e571a09af16ad2dab4b5e2c): Addressed @EdDev's comments.

orelmisan commented:
/retest-required

orelmisan commented:
/retest-required

EdDev commented on pkg/virt-controller/services/template.go:
```diff
@@ -258,24 +260,23 @@ func (t *templateService) RenderLaunchManifestNoVm(vmi *v1.VirtualMachineInstanc
-		multusNetworksAnnotation, err := multus.GenerateCNIAnnotationFromNameScheme(
-			vmi.Namespace, vmi.Spec.Domain.Devices.Interfaces, vmi.Spec.Networks, ordinalNameScheme, t.clusterConfig)
+	if t.netAnnotationsConverter != nil {
+		convertedNetAnnotations, err := t.netAnnotationsConverter.Convert(vmi, sourcePod)
```
I do not understand what this converts exactly. To me it just creates the annotation based on information coming from the VMI and pod.

I also think it does not fit well with the pattern used for other dependency injections as it is too specific.
Why it is not part of the annotations generators interface? After all, it is also an annotation generation action, just with an additional input, fitting a specific location (preparing a migration target pod).

What do you say about one of these options?
- Add another method for generating the annotations based on another input (pod).
- Take this out of here and move it to the caller (the migration controller), having a hook there for network manipulation on the target pod.

orelmisan commented on pkg/virt-controller/services/template_test.go:
```diff
@@ -5100,6 +5101,175 @@ var _ = Describe("Template", func() {
+		)
+
+		var (
+			vmi *v1.VirtualMachineInstance
```
Done.

orelmisan commented on pkg/virt-controller/services/template_test.go:
```diff
@@ -5100,6 +5101,175 @@ var _ = Describe("Template", func() {
+		})
+
+		It("Should call all registered annotation generators", func() {
+			generator1 := stubAnnotationsGenerator{
+				annotations: generator1Annotations,
+			}
+			generator2 := stubAnnotationsGenerator{
+				annotations: generator2Annotations,
+			}
```
Done.

orelmisan commented on pkg/virt-controller/services/template_test.go:
```diff
@@ -5100,6 +5101,175 @@ var _ = Describe("Template", func() {
+			Expect(pod.Annotations).To(HaveKeyWithValue(generator2Key2, generator2Value2))
+		})
+
+		DescribeTable("Should fail when the first generator failure is encountered", func(generator1, generator2 stubAnnotationsGenerator, expectedErr error) {
```
Since `WithAnnotationsGenerators` is a varadiac function that takes in the interface type:
```go
func WithAnnotationsGenerators(generators ...annotationsGenerator) templateServiceOption
```

It is not possible to pass it `[]stubAnnotationsGenerator` directly:
```go
WithAnnotationsGenerators(generators...)
```

It seems that we would need to convert the `[]stubAnnotationsGenerator` to`[]annotationsGenerator` first - which will just add more code and wouldn't provide any additional value.

orelmisan commented on pkg/virt-controller/services/template_test.go:
```diff
@@ -5100,6 +5101,175 @@ var _ = Describe("Template", func() {
+			_, err := svc.RenderLaunchManifest(vmi)
+			Expect(err).To(MatchError(expectedErr))
+		},
+			Entry("Generator1 fails",
```
Makes sense.
Removed the "When the first generator fails" scenario.

orelmisan commented on pkg/virt-controller/services/template_test.go:
```diff
@@ -5100,6 +5101,175 @@ var _ = Describe("Template", func() {
+
+		It("Should set a shared key's value by the last generator if two generators share a key", func() {
+			const (
+				sharedKey                = "sharedKey"
+				generator1SharedKeyValue = "sharedKeyValue1"
+				generator2SharedKeyValue = "sharedKeyValue2"
```
Done.

orelmisan commented on pkg/virt-controller/services/template_test.go:
```diff
@@ -5100,6 +5101,175 @@ var _ = Describe("Template", func() {
+			),
+		)
+
+		It("Should set a shared key's value by the last generator if two generators share a key", func() {
```
I've tried to follow Ginkgo's convention of It("should...").
Clearly it doesn't work well here.

Done.

orelmisan commented on pkg/virt-controller/services/template_test.go:
```diff
@@ -5100,6 +5101,175 @@ var _ = Describe("Template", func() {
+				generator2SharedKeyValue = "sharedKeyValue2"
+			)
+
+			generator1 := stubAnnotationsGenerator{
+				annotations: generator1Annotations,
+			}
+			generator1.annotations[sharedKey] = generator1SharedKeyValue
```
Removed the non-relevant annotations.

orelmisan commented:
[Change](https://github.com/kubevirt/kubevirt/compare/8b0d3c8de4684f7c0e571a09af16ad2dab4b5e2c..2eb1a80792fef3b4993b13aa222a1cfa9e7d10ba): Addressed @EdDev's comments.

orelmisan commented:
[Change](https://github.com/kubevirt/kubevirt/compare/2eb1a80792fef3b4993b13aa222a1cfa9e7d10ba..7edc7aac78b3997e044cac0e24a2d95ec9ee4955): Dropped the `netAnnotationsConverter` interface in favor of the more generic `targetAnnotationsGenerator`.

Dropped the `Converter` struct and incorporated its logic in `Generator`.

@EdDev PTAL.

EdDev reviewed (approved):
Thank you!

/approve

We will need another approver for the non-network changes.

enp0s3 commented on pkg/virt-controller/services/template.go:
```diff
@@ -258,24 +260,23 @@ func (t *templateService) RenderLaunchManifestNoVm(vmi *v1.VirtualMachineInstanc
-		ordinalNameScheme := namescheme.CreateOrdinalNetworkNameScheme(vmi.Spec.Networks)
-		multusNetworksAnnotation, err := multus.GenerateCNIAnnotationFromNameScheme(
-			vmi.Namespace, vmi.Spec.Domain.Devices.Interfaces, vmi.Spec.Networks, ordinalNameScheme, t.clusterConfig)
+	if t.netTargetAnnotationsGenerator != nil {
```
@orelmisan Why do you need to check this? isn't it mandatory for this generator to be registered as part of `vca.templateService = services.NewTemplateService(...`?

enp0s3 replied:
IIUC the migration controller is initialized with the template service instance, but the instance is already initialized and the generators were already registered via the template service init code:
```

	for _, opt := range opts {
		opt(&svc)
	}

```

orelmisan replied:
Thank you for reviewing this PR @enp0s3.

Since multiple unit tests use the concrete `templateService`, this condition is needed so these tests will not have to account for the network target annotations generator.

I think that our goal should be to eliminate this condition in the long run, which would require us to stub the templateService in multiple unit tests.

For completeness, the tests that use the concrete templateService are:
- pkg/storage/export/export/export_test.go
- pkg/storage/export/export/pvc-source_test.go
- pkg/storage/export/export/vm-source_test.go
- pkg/storage/export/export/vmsnapshot-source_test.go
- pkg/virt-controller/watch/application_test.go
- pkg/virt-controller/watch/migration_test.go
- pkg/virt-controller/watch/vmi_test.go

orelmisan replied:
We went with the same approch here for the same reason:
https://github.com/kubevirt/kubevirt/blob/097cc2a4538e21a628f69914da9ca2f629e891f2/pkg/virt-controller/services/template.go#L1459

EdDev replied:
@enp0s3 , I agree the condition is ugly and very confusing.
There are only two options I can think of to overcome it:
- Make the hook-point a slice and then just iterate over. [This was not done because there are no other potential clients to this at this location]
- Initialize a noop function on construction before calling  the option initializers (i.e. defaults).

We could have the last option implemented here if you will support it. It will be checked implicitly by the existing tests that do not add the option explicitly.

enp0s3 replied:
@EdDev Thank you for the clarification, 2nd option sounds good to me, anyway IMO we can do it in a follow up PR.

enp0s3 reviewed (commented):
@orelmisan Thank you! code looks great, I have a small question below

enp0s3 commented:
/approve
